airflow/providers/apache/hdfs/log/webhdfs_task_handler.py

```diff
@@ -0,0 +1,186 @@
+        is_trigger_log_context = getattr(ti, "is_trigger_log_context", False)
+        self.upload_on_close = is_trigger_log_context or not ti.raw
+
+        # Clear the file first so that duplicate data is not uploaded
+        # when re-using the same path (e.g. with rescheduled sensors)
+        if self.upload_on_close:
+            with open(self.handler.baseFilename, "w"):  # type: ignore[union-attr]
+                pass
+
+    @staticmethod
+    def encode_remote_path(remote_loc: str) -> str:
+        """
+        Encode URLs for compatibility with WebHDFS.
+
+        This will replace all ":" and "+" with "-".
+        """
+        # We replace all URL-encode characters with hyphen because:
+        # - ":" is not a valid character in HDFS paths.
+        # - "+" is present in timezones from many template variables
+        # - WebHDFS does not handle urlencoded correctly between versions:
+        #   - https://issues.apache.org/jira/browse/HDFS-14423
+        #   - https://issues.apache.org/jira/browse/HDFS-14466
+        # """
+        return remote_loc.replace(":", "-").replace("+", "-")
+
+    def close(self) -> None:
+        """Close and upload local log file to remote storage GCS."""
+        # When application exit, system shuts down all handlers by
+        # calling close method. Here we check if logger is already
+        # closed to prevent uploading the log to remote storage multiple
+        # times when `logging.shutdown` is called.
+        if self.closed:
+            return
+
+        super().close()
+
+        if not self.upload_on_close:
+            return
+
+        local_loc = os.path.join(self.local_base, self.log_relative_path)
+        remote_loc = os.path.join(self.remote_base, self.encode_remote_path(self.log_relative_path))
+
+        if os.path.exists(local_loc):
+            log = pathlib.Path(local_loc).read_text()
+            self.hdfs_write(log, remote_loc)
+
+        # Mark closed, so we don't double write if close is called twice
+        self.closed = True
+
+    def _read_remote_logs(
+        self, ti: TaskInstance, try_number: int, metadata: dict[str, Any] | None = None
+    ) -> tuple[list[str], list[str]]:
+        # Explicitly getting log relative path is necessary as the given
+        # task instance might be different than task instance passed in
+        # in set_context method.
+        worker_log_rel_path = self._render_filename(ti, try_number)
+
+        remote_loc = os.path.join(self.remote_base, worker_log_rel_path)
+        remote_loc = self.encode_remote_path(remote_loc)
+
+        logs = []
+        messages = []
+        if self.hook.check_for_path(remote_loc):
+            messages.append(f"Found logs in webhdfs: {remote_loc}")
+            try:
+                messages.append(f"Reading remote log from HDFS: {remote_loc}.")
+                logs.extend(self.hook.read_file(remote_loc).splitlines())
```

Comment:
```suggestion
                logs.append(self.hook.read_file(remote_loc))
```
the `logs` object returned should be one element per file